### ISSUE TITLE
Implement own bilinear form

### ISSUE THREAD

**Issue by YHDAUT** [mfem/mfem#4104]
Hi,  it seems to me not trivial to implement the mixed bilinear form :

 \partial_x u \partial_x v + \partial_y u \partial_y v - \partial_z u \partial_z v 

I am trying to make this new bilinear form from the standard 
diffusion integrator in MFEM.

However, is there any way/example to show how to 
select derivative in each direction x, y, z

\partial_x u \partial_x v 
\partial_y u \partial_y v
\partial_z u \partial_z v

Then perform summation or subtraction of these terms.

I found 

MixedScalarWeakDerivativeIntegrator 

which seems to compute derivative in 1D,

but not sure if this is the derivative that can be used for compute 
my particular bilinear form.

Thanks

The original diffusion operator (below) performs  summation
 \partial_x u \partial_x v + \partial_y u \partial_y v + \partial_z u \partial_z v 
instead of subtraction of the last term.  

//original mfem diffusion operator: 

void DiffusionIntegrator::AssembleElementMatrix2(
   const FiniteElement &trial_fe, const FiniteElement &test_fe,
   ElementTransformation &Trans, DenseMatrix &elmat)
{
   int tr_nd = trial_fe.GetDof();
   int te_nd = test_fe.GetDof();
   dim = trial_fe.GetDim();
   int spaceDim = Trans.GetSpaceDim();
   bool square = (dim == spaceDim);
   double w;

   if (VQ)
   {
      MFEM_VERIFY(VQ->GetVDim() == spaceDim,
                  "Unexpected dimension for VectorCoefficient");
   }
   if (MQ)
   {
      MFEM_VERIFY(MQ->GetWidth() == spaceDim,
                  "Unexpected width for MatrixCoefficient");
      MFEM_VERIFY(MQ->GetHeight() == spaceDim,
                  "Unexpected height for MatrixCoefficient");
   }

#ifdef MFEM_THREAD_SAFE
   DenseMatrix dshape(tr_nd, dim), dshapedxt(tr_nd, spaceDim);
   DenseMatrix te_dshape(te_nd, dim), te_dshapedxt(te_nd, spaceDim);
   DenseMatrix invdfdx(dim, spaceDim);
   DenseMatrix dshapedxt_m(te_nd, MQ ? spaceDim : 0);
   DenseMatrix M(MQ ? spaceDim : 0);
   Vector D(VQ ? VQ->GetVDim() : 0);
#else
   dshape.SetSize(tr_nd, dim);
   dshapedxt.SetSize(tr_nd, spaceDim);
   te_dshape.SetSize(te_nd, dim);
   te_dshapedxt.SetSize(te_nd, spaceDim);
   invdfdx.SetSize(dim, spaceDim);
   dshapedxt_m.SetSize(te_nd, MQ ? spaceDim : 0);
   M.SetSize(MQ ? spaceDim : 0);
   D.SetSize(VQ ? VQ->GetVDim() : 0);
#endif
   elmat.SetSize(te_nd, tr_nd);

   const IntegrationRule *ir = IntRule ? IntRule : &GetRule(trial_fe, test_fe);

   elmat = 0.0;
   for (int i = 0; i < ir->GetNPoints(); i++)
   {
      const IntegrationPoint &ip = ir->IntPoint(i);
      trial_fe.CalcDShape(ip, dshape);
      test_fe.CalcDShape(ip, te_dshape);

      Trans.SetIntPoint(&ip);
      CalcAdjugate(Trans.Jacobian(), invdfdx);
      w = Trans.Weight();
      w = ip.weight / (square ? w : w*w*w);
      Mult(dshape, invdfdx, dshapedxt);
      Mult(te_dshape, invdfdx, te_dshapedxt);
      // invdfdx, dshape, and te_dshape no longer needed
      if (MQ)
      {
         MQ->Eval(M, Trans, ip);
         M *= w;
         Mult(te_dshapedxt, M, dshapedxt_m);
         AddMultABt(dshapedxt_m, dshapedxt, elmat);
      }
      else if (VQ)
      {
         VQ->Eval(D, Trans, ip);
         D *= w;
         AddMultADAt(dshapedxt, D, elmat);
      }
      else
      {
         if (Q)
         {
            w *= Q->Eval(Trans, ip);
         }
         dshapedxt *= w;
         AddMultABt(te_dshapedxt, dshapedxt, elmat);
      }
   }
}

_Originally posted by @YHDAUT in https://github.com/mfem/mfem/issues/4094#issuecomment-1920099749_

**Comment by atallah727** [comment]
Hello @YHDAUT, you can implement a mixed bilinear form you can check ex5p which assembles and solves the Darcy equations.

**Comment by YHDAUT** [comment]
Hi, thanks for pointing out this ex5p. I am actually looking for the implementation of 
a kind of diffusion operator:

\partial_x u \partial_x v **+**
 \partial_y u \partial_y v **-** 
\partial_z u \partial_z v

which is not implemented in MFEM.  I am wondering if it is possible to 
implement this new integrator by modifying the original diffusion 
integrator operator in the MFEM

void DiffusionIntegrator::AssembleElementMatrix2(
const FiniteElement &trial_fe, const FiniteElement &test_fe,
ElementTransformation &Trans, DenseMatrix &elmat)
{
 ........
}

Would it be possible to provide some  more details on how to 
make this new integrator ... 

Thanks

**Comment by atallah727** [comment]
Hello @YHDAUT, the most straight forward thing to do is just write nested loops over `dofs` and `dim` for both trial and test function spaces and assemble to `elmat` by taking the product of components of the gradients through the indices. For example,
```
// assembly of (grad u, grad w)
for (int i = 0; i < dof; i++)
   {
       for (int j = 0; j < dof; j++)
          {
             for (int k = 0; k < dim; k++)
                {
                   elmat(i, j) += dshape_ps(i,k) * dshape_ps(j,k) * w;
                 }
          }
   }
```
The above code assembles the diffusion operator through indices. You could do the same for your operator.

**Comment by YHDAUT** [comment]
Hi @tzanio  @atallah727 .  many thanks for this great help.   Following your instructions,  I have made the following code for my integrator:

ParMixedBilinearForm *bVarf(new ParMixedBilinearForm(ufespace, pfespace));
bVarf->AddDomainIntegrator(new MyDiffusionIntegrator);

//

class MyDiffusionIntegrator: public BilinearFormIntegrator
{

private:
   Vector vec, vecdxt, pointflux, shape;
#ifndef MFEM_THREAD_SAFE
   DenseMatrix dshape, dshapedxt, invdfdx, M, dshapedxt_m;
   DenseMatrix te_dshape, te_dshapedxt;
   Vector D;
#endif

   // PA extension
   const FiniteElementSpace *fespace;
   const DofToQuad *maps;         ///< Not owned
   const GeometricFactors *geom;  ///< Not owned
   int dim, ne, dofs1D, quad1D;
   Vector pa_data;

public:
   /// Construct a diffusion integrator wit  h coefficient Q = 1
   MyDiffusionIntegrator(const IntegrationRule *ir = nullptr)
      : BilinearFormIntegrator(ir),
        Q(NULL), VQ(NULL), MQ(NULL), SMQ(NULL), maps(NULL), geom(NULL) { }


   /** Given a particular Finite Element computes the element stiffness matrix
       elmat. */
   virtual void AssembleElementMatrix(const FiniteElement &el,
                                      ElementTransformation &Trans,
                                      DenseMatrix &elmat){;}
   /** Given a trial and test Finite Element computes the element stiffness
       matrix elmat. */
   virtual void AssembleElementMatrix2(const FiniteElement &trial_fe,
                                       const FiniteElement &test_fe,
                                       ElementTransformation &Trans,
                                       DenseMatrix &elmat);

  
   using BilinearFormIntegrator::AssemblePA;

   static const IntegrationRule &GetRule(const FiniteElement &trial_fe,
                                         const FiniteElement &test_fe);

   bool SupportsCeed() const { return DeviceCanUseCeed(); }
};


void MyDiffusionIntegrator::AssembleElementMatrix2(
   const FiniteElement &trial_fe, const FiniteElement &test_fe,
   ElementTransformation &Trans, DenseMatrix &elmat)
{
   int tr_nd = trial_fe.GetDof();
   int te_nd = test_fe.GetDof();
   dim = trial_fe.GetDim();

#ifdef MFEM_THREAD_SAFE
   DenseMatrix dshape(tr_nd, dim), dshapedxt(tr_nd, spaceDim);
   DenseMatrix te_dshape(te_nd, dim), te_dshapedxt(te_nd, spaceDim);
   DenseMatrix invdfdx(dim, spaceDim); 
#else
   dshape.SetSize(tr_nd, dim);
   dshapedxt.SetSize(tr_nd, spaceDim);
   te_dshape.SetSize(te_nd, dim);
   te_dshapedxt.SetSize(te_nd, spaceDim);
   invdfdx.SetSize(dim, spaceDim);
#endif
   elmat.SetSize(te_nd, tr_nd);

   const IntegrationRule *ir = IntRule ? IntRule : &GetRule(trial_fe, test_fe);

   elmat = 0.0;
   for (int i = 0; i < ir->GetNPoints(); i++)
   {
      const IntegrationPoint &ip = ir->IntPoint(i);
      trial_fe.CalcDShape(ip, dshape);
      test_fe.CalcDShape(ip, te_dshape);

      Trans.SetIntPoint(&ip);
      CalcAdjugate(Trans.Jacobian(), invdfdx);
      w = ip.weight / Trans.Weight();
      Mult(dshape, invdfdx, dshapedxt);
      Mult(te_dshape, invdfdx, te_dshapedxt);

      for (int i = 0; i < te_nd; i++)
         {
            for (int j = 0; j < tr_nd; j++)
              {
                for (int k = 0; k < dim-1; k++)
                  {
                    elmat(i, j) += dshape_ps(i,k) * dshape_ps(j,k) * w;
                  }
                  elmat(i, j) -= dshape_ps(i,dim-1) * dshape_ps(j,dim-1) * w;
              }
          }
   }
}

const IntegrationRule &MyDiffusionIntegrator::GetRule(
   const FiniteElement &trial_fe, const FiniteElement &test_fe)
{
   int order;
   if (trial_fe.Space() == FunctionSpace::Pk)
   {
      order = trial_fe.GetOrder() + test_fe.GetOrder() - 2;
   }
   else
   {
      // order = 2*el.GetOrder() - 2;  // <-- this seems to work fine too
      order = trial_fe.GetOrder() + test_fe.GetOrder() + trial_fe.GetDim() - 1;
   }

   if (trial_fe.Space() == FunctionSpace::rQk)
   {
      return RefinedIntRules.Get(trial_fe.GetGeomType(), order);
   }
   return IntRules.Get(trial_fe.GetGeomType(), order);
}

//

I have still some concerns:  

1.  Do I have to overload AssembleElementMatrix?  I have only considered mixed form. Is it OK if I only overload the AssembleElementMatrix2

2. Will this code be different in parallel and sequential versions? Shall we change something in the code if we switch to the parallel implementations? 

Thanks a lot for your time and patience.

**Comment by atallah727** [comment]
Hello @YHDAUT: 
1. You need to overload `AssembleElementMatrix ` when you are integrating a term involving a trial function for a some solution variable, let's say `u`,  and its corresponding test function. In the example I gave you, because it is a diffusion integrator  $`(\nabla u, \nabla \tilde{u})_{\Omega} \! `$   the trial and test correspond to the same solution variable so I only had to change `AssembleElementMatrix`. On the other hand, when you are solving the Stokes or Darcy pde then you would end up with "cross" terms such as ( \div u ,

**Comment by YHDAUT** [comment]
Hello @atallah727 . Many thanks for your kind help. 

Yes, since  I have a cross term,  I will only need to overload the AssembleElementMatrix2 function

I will pay attention to the definition with Par prefix.  In the above AssembleElementMatrix2, it seems to me there are no variables that I need to put Par prefix  in MyDiffusionIntegrator class if my ufespace and pfespace are parallel finite element spaces:

ParMixedBilinearForm *bVarf(new ParMixedBilinearForm(ufespace, pfespace));
bVarf->AddDomainIntegrator(new MyDiffusionIntegrator);

although in MyDiffusionIntegrator class,  there is a varaible defined as 

const FiniteElementSpace *fespace

I hope I understand your lines correctly. 

Thanks

**Comment by atallah727** [comment]
Yes, make sure that you use `const ParFiniteElementSpace *fespace` instead of `const FiniteElementSpace *fespace`. There are communication calls for variables defined with the `Par` prefix.

**Comment by YHDAUT** [comment]
Thank you @atallah727 .  Do you mean that I have to 
adapt my diffusion integrator so that inside the arguments 
in AssembleElementMatrix2 shall be ParFiniteElement ?

class MyDiffusionIntegrator: public BilinearFormIntegrator
{

virtual void AssembleElementMatrix2(**const ParFiniteElement** &trial_fe,
**const ParFiniteElement** &test_fe,
ElementTransformation &Trans,
DenseMatrix &elmat);
};

instead of 


class MyDiffusionIntegrator: public BilinearFormIntegrator
{

virtual void AssembleElementMatrix2(const FiniteElement &trial_fe,
const FiniteElement &test_fe,
ElementTransformation &Trans,
DenseMatrix &elmat);
};

I may misunderstand.  I thought MyDiffusionIntegrator class is the same for 
both sequential and parallel codes. 

Many thanks.

**Comment by atallah727** [comment]
There is no `ParFIniteElement`, `FiniteElement` is for serial and parallel. I was responding to your previous comment on `FiniteElementSpace` where I said it should be `ParFiniteElementSpace`.

**Comment by YHDAUT** [comment]
Thanks a lot.
